steamtinkerlaunch

```diff
@@ -25924,7 +25929,13 @@ function checkSteamDeckDependencies {
 		local CHECKCMD
 		CHECKCMD="$($DEPCMD &> /dev/null --version && echo "OK" || echo "NOK")"
 
-		if [ -f "$(command -v "$DEPCMD")" ] && [ "$CHECKCMD" = "OK" ]; then
+
+		# The check now says, don't update dependencies if all of these conditions are true:
+		#  1. The dependency file exists
+		#  2. The dependency can actually be ran, confirming it is a valid file
+		#  3. SteamTinkerLaunch has not updated or autoupdater isn't enabled, so there would be no change in dependency version and thus no need to update
+		# If any of these are false, we need to check our dependencies (if a file is missing we would need to update, or if it cannot be used we need to update, and also if STL updated we may need a newer version, so update).
+		if [[ -f "$(command -v "$DEPCMD")" && "$CHECKCMD" = "OK" && ( "$( checkSteamDeckSTLUpdated )" -eq 0 || "$STEAMDECK_AUTOUP" -eq 0 ) ]]; then
```

Comment:
With the changes  to how `checkSteamDeckSTLUpdated` works, it'll return `0` (positive, meaning an update occurred), so we can use it like a true boolean value. Note how it can be used without square brackets, because it's not a `test` condition (Bash is weird).

```suggestion
		if [[ -f "$(command -v "$DEPCMD")" && "$CHECKCMD" = "OK" && ( checkSteamDeckSTLUpdated || "$STEAMDECK_AUTOUP" -eq 0 ) ]]; then
```

ShellCheck also shouldn't complain about this, with the code below at least it was happy.

<hr>

I verified this sort of logic works with a little test script, hopefully it helps illustrate the logic a little better :-)

```bash
#!/usr/bin/bash

PROGVERS="1.2.3"

function hasUpdated {
    oldvers="1..3"
    
    if [ "${oldvers}" = "${PROGVERS}" ]; then
        # no update
        return 1
    else
        return 0
    fi
}

boolvar=1  # Set this to 0 to get "no update"
if ( hasUpdated && [ "${boolvar}" -eq 1 ] ); then
    echo "Has updated"
else
    echo "no update"
fi
```

Reply:
Was about to fall asleep and it just dawned on me that this should probably be a `! checkSteamDeckSTLUpdated` check, since before we were checking `-eq 0`. 

I think there is some weirdness around how Bash treats the `!` negation, so please test if you can to make sure the negation doesn't affect the whole statement and only that `checkSteamDeckSTLUpdated` check.

I'll try do some testing in a test script tomorrow to see how Bash handles this. I don't know if there's any examples in the codebase of this kind of check in a compound `if` check.

Reply:
I just tried this solution and came across an issue, it doesn't look like the return value of a function is supported inside a test `[[`. See https://stackoverflow.com/a/42971383 
I'm not sure how we should tweak it, one idea could be to "preload" the function response before the answer output

Reply:
> I just tried this solution and came across an issue, it doesn't look like the return value of a function is supported inside a test

That's a bit strange, is the code snippet I provided invalid then? It seemed to work in my testing...

Plus, we use `if funcThatReturnsZero ; then` syntax around the codebase already.

Reply:
It's not invalid, in the current `if` it's more like ˋif [[ hasUpdated && ( "${boolvar}" -eq 1 ) ]]` and I don't know how to properly wrap the `or` to override the update if they're disabled

Reply:
Would this work?

```
[[ -f "$(command -v "gamescope")" && "$CHECKCMD" = "OK" ]] && ( checkSteamDeckSTLUpdated || [[ "$STEAMDECK_AUTOUP" -eq 0 ]] ); then
    # ...
fi
```

I'm playing around with this in a test script. The logic is a little complex here, but I think this should work, since `( checkSteamDeckSTLUpdated )` should capture the return, and we wrap the check on `$STEAMDECK_AUTOUP` in a separate `[[` test.

Reply:
Let me write a truth table, I think it will be more clear for the `If STEAMDECK_AUTOUP is 1, we will always need to update regardless of DEPCMD, CHECKCMD, and checkSteamDeckSTLUpdated` part.

Variables | `DEPCMD` | `CHECKCMD` | `!checkSteamDeckSTLUpdated` | `STEAMDECK_AUTOUP` | Do we update ? | If line boolean value
--- | --- | --- | --- |--- |--- |---
| 0 | 0 | 0 | 0 | 0 | YES | 0 |
| 1 | 0 | 0 | 0 | 1 | YES | 0 |
| 2 | 0 | 0 | 1 | 0 | YES | 0 |
| 3 | 0 | 0 | 1 | 1 | YES | 0 |
| 4 | 0 | 1 | 0 | 0 | YES | 0 |
| 5 | 0 | 1 | 0 | 1 | YES | 0 |
| 6 | 0 | 1 | 1 | 0 | YES | 0 |
| 7 | 0 | 1 | 1 | 1 | YES | 0 |
| 8 | 1 | 0 | 0 | 0 | YES | 0 |
| 9 | 1 | 0 | 0 | 1 | YES | 0 |
| 10 | 1 | 0 | 1 | 0 | YES | 0 |
| 11 | 1 | 0 | 1 | 1 | YES | 0 |
| 12 | 1 | 1 | 0 | 0 | NO (no update and autoupdater not enabled) | 1 |
| 13 | 1 | 1 | 0 | 1 | NO (no update and autoupdater enabled) | 1 |
| 14 | 1 | 1 | 1 | 0 | NO (there's an update but autoupdater isn't enabled) | 1 |
| 15 | 1 | 1 | 1 | 1 | YES (there's an update and autoupdater enabled)  | 0 |

Notice how the 12 first lines basically describe the initial `if`, we don't want to tamper those because tampering those would mean we could be in a case where the user has no libs anymore or non-working ones.
The autoupdate strategy appears at the end with the new additional `AND`. 

I think I made a mistake initially, the right thing to add is `AND not ( !checkSteamDeckSTLUpdated && STEAMDECK_AUTOUP -eq 1)` instead. Bash logical operators and 0 doesn't help

Feel free to correct me if I made a mistake

Reply:
It looks like the error we got before :  `steamtinkerlaunch: line 25938: [: : integer expression expected` comes from the fact that the new env variable `STEAMDECK_AUTOUP` is not populated inside `global.conf` after an update. How would I fix that ?

The if line looks to work with `if [[ -f "$(command -v "$DEPCMD")" && "$CHECKCMD" = "OK" ]] && ! ( checkSteamDeckSTLUpdated && [ "$STEAMDECK_AUTOUP" -eq 1 ] ); then` :tada:

Reply:
The config file values get updated each time there is a version bump, so if you make a new field the version needs changed. It doesn't need to be exact, the string just needs to change to something unique I believe :-)

Reply:
That truth table is really awesome by the way. Makes the reasoning for the conditional you came up with a lot more sense. I think this comment is resolved!